lens/fun_dep.ml

```diff
@@ -98,25 +173,142 @@ module Tree = struct
+      |> List.map ~f:left
+      |> List.sort_uniq Alias.Set.compare
+    in
+    let remain =
+      Alias.Set.diff key
+        (List.fold_right Alias.Set.union subkeys Alias.Set.empty)
+    in
+    let of_sub_key key =
+      let subfds = Set.filter (left >> Alias.Set.equal key) fds in
+      let remaining = Set.diff fds subfds in
+      let subfds =
+        List.map ~f:(right >> fd_subnodes remaining) (Set.elements subfds)
+        |> List.flatten
+      in
+      FDNode (key, subfds)
+    in
+    let subtrees = List.map ~f:of_sub_key subkeys in
+    if Alias.Set.is_empty remain then subtrees
+    else FDNode (remain, []) :: subtrees
+
+  let is_disjoint tree ~columns =
+    (* for a tree to be disjoint, traverse a tree and ensure that the columns of
+       each node are in columns, then continue traversing the tree without those columns. *)
+    let columns = ref columns in
+    let joint = ref Alias.Set.empty in
+    let rec traverse (FDNode (c, subnodes)) =
+      if Alias.Set.subset c !columns then (
+        columns := Alias.Set.diff !columns c ;
+        List.for_all ~f:traverse subnodes )
+      else (
+        joint := c ;
+        false )
+    in
+    List.for_all ~f:traverse tree
+    |> Result.of_bool ~error:(Check_error.FunDepNotTreeForm !joint)
+
+  let rec all_cols tree =
+    let f tree =
+      let sub = subnodes tree |> all_cols in
+      let node = cols tree in
+      Alias.Set.union sub node
+    in
+    let l = List.map ~f tree in
+    Alias.Set.union_all l
+
+  let of_fds fds ~columns =
+    let roots = Set.simple_key fds ~columns in
+    let tree = fd_subnodes fds roots in
+    let not_included = Alias.Set.diff (Set.all_columns fds) (all_cols tree) in
+    Alias.Set.is_empty not_included
+    |> Result.of_bool ~error:(Check_error.ProbablyCycle not_included)
+    >>= fun () -> is_disjoint tree ~columns >>| fun () -> tree
+
+  let split_fds fds colsets =
+    let split_fd fd =
+      let rights =
+        Alias.Set.Set.filter (fun v -> Alias.Set.subset v (right fd)) colsets
+      in
+      let l = left fd in
+      let unmapped =
+        Alias.Set.Set.elements rights
+        |> Alias.Set.union_all
+        |> Alias.Set.diff (right fd)
+      in
+      let splitto = Alias.Set.Set.elements rights |> List.map ~f:(make l) in
+      if Alias.Set.is_empty unmapped then splitto
+      else make l unmapped :: splitto
+    in
+    Set.elements fds |> List.map ~f:split_fd |> List.flatten |> Set.of_list
+
+  module Tree_form_error = struct
+    type t = ContainsCycle of Alias.Set.t list | NotDisjoint of Alias.Set.t
+    [@@deriving eq, show]
+
+    exception E of t
+  end
+
+  let is_acyclic fds =
+    let nextfds node =
+      Set.filter (fun fd -> Alias.Set.equal (left fd) node) fds
+    in
+    let rec find_cycle node bt =
+      match bt with
+      | x :: xs ->
+          if Alias.Set.equal x node then List.append bt [node]
+          else find_cycle node xs
+      | [] -> [Alias.Set.singleton "ERROR RECONSTRUCTING CYCLE"]
```

Comment:
This seems a bit of a hack -- could it lead to "ERROR RECONSTRUCTING CYCLE" ending up in a user program? I'd expect an exception would be a better way to handle this

Reply:
If a cycle is found, the function `find_cycle` tries to reconstruct the cycle. Hypothetically this should always correctly terminate as it is used to produce an exception. I don't really want exception handling code to produce more exceptions.

Reply:
OK - in which case, could you make this either an empty set or assert false, depending on the logic?